### PR TITLE
Allow blocks in deployment params

### PR DESCRIPTION
**Pull request by GalLadislav**
### Context

Related #14393

I tried to "fix" Prefect's 3.0 inability to pass Blocks as deployments parameters. This feature is really important for me personally (and probably for others too). Since the related issue was closed as "not planned", i implemented something that could provide at least some functionality until it is properly addressed. 

I would appreciate a feedback if this PR is useful and is worthwhile to polish it and add unittests etc. Thank you in advance for any guidance.

### Implementation

Implementation is building on already existing UI features in FlowRunCreate for deployment runs that use passing `{"$ref": "__uuid__"}` as block reference. I have tracked to main issues which causes this feature to not work. 
1. Server's validation API does not allow this block definition,
2. `Block` does not processes the **$ref** in init kwargs.

#### Server validation API

To allow **$ref** for blocks, preprocessing schemas modify block definition to allow either block definition or block reference.
Added two **$ref** formats, one that supports deployment flow run create form and one that supports multiple reference types.

#### Block initialization

For Block class i added method `Block.load_from_ref` that process either block document ID or supported reference dictionary format and returns loaded block with data from references block document.
Other changes:
- created `Block._load_from_block_document` method to deal with duplicate code used both in `Block.load` and `Block.load_from_ref`.
- added `Block._get_block_document_by_id`
- `Flow.validate_parameters` updated to transform any block reference in given parameters to corresponding block with `Block.load_from_ref`.

#### Issues

- API validation might check if blocks exist

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`.
- [x] If this pull request adds new functionality, it includes unit tests that cover the changes
- [x] If this pull request adds functions or classes, it includes helpful docstrings.

### CODE REVIEW

File: src/prefect/blocks/core.py
Comment (cicdw) [inline src/prefect/blocks/core.py:808]: we are still supporting Python 3.9 and so can't support this syntax quite yet! You'll need to use `typing.Union` here

Comment (GalLadislav) [reply src/prefect/blocks/core.py]: Thank you, fixed

File: src/prefect/blocks/core.py
Comment (cicdw) [inline src/prefect/blocks/core.py:817]:
```suggestion
                )
```
no need to add more to the traceback here

File: src/prefect/blocks/core.py
Comment (cicdw) [inline src/prefect/blocks/core.py:826]:
```suggestion
            )
```

File: src/prefect/blocks/core.py
Comment (cicdw) [inline src/prefect/blocks/core.py:278]: because this is _always_ run synchronously, I suggest adding `_sync=True` to this call to ensure it's called properly.

**Review by cicdw — CHANGES_REQUESTED** [review]
a few code nit picks - also could you add a test for what this initialization looks like, and also a test for using a block type hint with a flow that validates its parameters?

Ultimately it'd be better (and easier to reason about) if the API could pass the block slug to the client for validation, but this might be an OK patch in the meantime as that will involve more changes.

I've also tagged @desertaxle for review here, as he might have more insight into this implementation and whether there are any edge cases we should consider.

**Comment by GalLadislav** [comment]
I added block reference validation to check if it matches initialized block type. Not sure if both slug and type should be compared, or only one of them or should be validated differently.
Added also some tests to check basic cases. Probably more should be added. Also I was not sure if the location is ok.

File: src/prefect/utilities/schema_tools/validation.py
Comment (desertaxle) [inline src/prefect/utilities/schema_tools/validation.py:274]: The reference format used when nested block documents in other block documents are `{"$ref": {"block_document_id": block_document_id } }.` I'd prefer to use that same format here for consistency and to leave room for other reference types in the future (like using the block slug as @cicdw suggested).

File: src/prefect/blocks/core.py
Comment (desertaxle) [inline src/prefect/blocks/core.py:280]: I'm wary of putting a blocking API call in `__init__`. Could you add a new method to load a block from an ID? That would require an update to `Flow.validate_parameters` to use that new method, but I think that'd make this new functionality more explicit.

**Review by desertaxle — CHANGES_REQUESTED** [review]
Thanks for the PR @GalLadislav, this will be a super useful feature! Let me know if you have any questions about any of my requested changes!

File: src/prefect/blocks/core.py
Comment (cicdw) [reply src/prefect/blocks/core.py:280]: Good call, agreed on this -- maybe it could be a classmethod `load_from_ref` that the `validate_parameters` method calls explicitly for Block typed parameters.

File: src/prefect/utilities/schema_tools/validation.py
Comment (GalLadislav) [reply src/prefect/utilities/schema_tools/validation.py:274]: Do i understand correctly that you request changes to the UI itself? i tried to fit current solution to current UI functions, since changes to UI would require more work. I can do that, but it would take more time.

File: src/prefect/blocks/core.py
Comment (GalLadislav) [reply src/prefect/blocks/core.py]: Refactored to new method `load_from_ref` which is used for preprocessing parameters in `validate_parameters`. To deal with duplicate code i dared to create `_load_from_block_document` method to be used both by `load` and by `load_from_ref`.

File: src/prefect/utilities/schema_tools/validation.py
Comment (GalLadislav) [reply src/prefect/utilities/schema_tools/validation.py]: Added additional schema to enable your proposed reference formats. I kept the current format, as it is used by Deployment Flow Run Create form. I hope i understood your request correctly.

File: tests/blocks/test_block_reference.py
Comment (cicdw) [inline tests/blocks/test_block_reference.py:24]: Could you write a test for `load_from_ref` that includes a nested block document? e.g., 
```python
class NestedBlock(Block):
    child: OtherReferencedBlock
```

**Review by cicdw — CHANGES_REQUESTED** [review]
This looks pretty solid to me! I requested one additional test for nested block references, and will wait on @desertaxle's final review as well. 

Thank you for taking the time to work on this @GalLadislav !!

File: tests/blocks/test_block_reference.py
Comment (GalLadislav) [reply tests/blocks/test_block_reference.py:24]: I am glad you noticed that these are missing. Added few more test for this including nesting with `pydantic.BaseModel` as flow parameter.

File: src/prefect/utilities/schema_tools/validation.py
Comment (desertaxle) [reply src/prefect/utilities/schema_tools/validation.py]: Thanks for adding that! I forgot that we'd need to update the UI to populate the backend data to match that format. I can create a separate issue to allow different reference shapes.

**Review by desertaxle — APPROVED** [review]
LGTM! Thanks for the awesome contribution @GalLadislav!

**Review by cicdw — APPROVED** [review]
Thanks again @GalLadislav for working with us on this and being so responsive to our feedback!!

**Comment by GalLadislav** [comment]
Thank you @desertaxle @cicdw for your time and guidance. I am happy i could contribute.

**Pull request merged**
